File: app/src/main/java/com/dumuzeyn/mp3player/LyricsRepository.java
package com.dumuzeyn.mp3player;

import android.content.ContentResolver;
import android.content.Context;
import android.net.Uri;
import android.provider.DocumentsContract;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.RejectedExecutionException;
import java.util.LinkedHashMap;
import java.util.Map;

/** Reads bounded local sidecar and embedded lyrics through ContentResolver only. */
final class LyricsRepository implements AutoCloseable {
    interface Callback {
        void loaded(LrcDocument document);
    }

    private static final int MAX_LYRICS_BYTES = 1024 * 1024;
    private final Context context;
    private final android.os.Handler mainHandler;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private volatile boolean closed;
    private final Map<String, LrcDocument> cache = new LinkedHashMap<String, LrcDocument>(
            8, 0.75f, true) {
        @Override protected boolean removeEldestEntry(Map.Entry<String, LrcDocument> eldest) {
            return size() > 12;
        }
    };

    LyricsRepository(Context context, android.os.Handler mainHandler) {
        this.context = context;
        this.mainHandler = mainHandler;
    }

    void load(Track track, Callback callback) {
        synchronized (cache) {
            LrcDocument cached = cache.get(track.uri);
            if (cached != null) {
                mainHandler.post(() -> callback.loaded(cached));
                return;
            }
        }
        try {
            executor.execute(() -> {
                LrcDocument result = find(track);
                synchronized (cache) {
                    cache.put(track.uri, result);
                }
                mainHandler.post(() -> {
                    if (!closed) {
                        callback.loaded(result);
                    }
                });
            });
        } catch (RejectedExecutionException ignored) {
            // Activity is already closing.
        }
    }

    @Override
    public void close() {
        closed = true;
        synchronized (cache) {
            cache.clear();
        }
        executor.shutdownNow();
    }

    private LrcDocument find(Track track) {
        for (Uri candidate : sidecars(track)) {
            String value = read(candidate);
            if (!value.trim().isEmpty()) {
                return new LrcParser().parse(value);
            }
        }
        try {
            String embedded = new EmbeddedLyricsReader().read(context, Uri.parse(track.uri));
            if (!embedded.trim().isEmpty()) {
                return new LrcParser().parse(embedded);
            }
        } catch (RuntimeException error) {
            VoltuneLog.failure("embedded_lyrics_uri_failed", error);
        }
        return new LrcDocument(new ArrayList<>(), "", false);
    }

    private List<Uri> sidecars(Track track) {
        ArrayList<Uri> result = new ArrayList<>();
        try {
            Uri source = Uri.parse(track.uri);
            if (!"content".equalsIgnoreCase(source.getScheme())
                    || !DocumentsContract.isDocumentUri(context, source)) {
                return result;
            }
            String id = DocumentsContract.getDocumentId(source);
            int slash = id.lastIndexOf('/');
            String parent = slash < 0 ? "" : id.substring(0, slash + 1);
            String filename = slash < 0 ? id : id.substring(slash + 1);
            int extension = filename.lastIndexOf('.');
            String base = extension > 0 ? filename.substring(0, extension) : filename;
            addCandidate(result, source, parent + base + ".lrc");
            addCandidate(result, source, parent + base + ".txt");
        } catch (RuntimeException ignored) {
            // A malformed or unsupported provider URI simply has no sidecar lyrics.
        }
        return result;
    }

    private void addCandidate(List<Uri> target, Uri source, String documentId) {
        try {
            Uri candidate = source.toString().contains("/tree/")
                    ? DocumentsContract.buildDocumentUriUsingTree(source, documentId)
                    : DocumentsContract.buildDocumentUri(source.getAuthority(), documentId);
            target.add(candidate);
        } catch (RuntimeException ignored) {
            // The provider does not support a sibling document URI.
        }
    }

    private String read(Uri uri) {
        ContentResolver resolver = context.getContentResolver();
        try (InputStream input = resolver.openInputStream(uri)) {
            if (input == null) {
                return "";
            }
            ByteArrayOutputStream output = new ByteArrayOutputStream();
            byte[] buffer = new byte[8192];
            int total = 0;
            int count;
            while ((count = input.read(buffer)) >= 0) {
                total += count;
                if (total > MAX_LYRICS_BYTES) {
                    return "";
                }
                output.write(buffer, 0, count);
            }
            String value = new String(output.toByteArray(), StandardCharsets.UTF_8);
            return value.startsWith("\ufeff") ? value.substring(1) : value;
        } catch (Exception ignored) {
            return "";
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/LyricsRepository.kt
package com.dumuzeyn.mp3player

import android.content.Context
import android.net.Uri
import android.os.Handler
import android.provider.DocumentsContract
import java.io.ByteArrayOutputStream
import java.util.LinkedHashMap
import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException

/** Reads bounded local sidecar and embedded lyrics through ContentResolver only. */
internal class LyricsRepository(
    private val context: Context,
    private val mainHandler: Handler,
) : AutoCloseable {
    fun interface Callback {
        fun loaded(document: LrcDocument)
    }

    private val executor = Executors.newSingleThreadExecutor()
    @Volatile private var closed = false
    private val cache = object : LinkedHashMap<String, LrcDocument>(8, 0.75f, true) {
        override fun removeEldestEntry(
            eldest: MutableMap.MutableEntry<String, LrcDocument>?,
        ): Boolean = size > MAX_CACHE_ENTRIES
    }

    fun load(track: Track, callback: Callback) {
        synchronized(cache) {
            cache[track.uri]?.let { cached ->
                mainHandler.post { callback.loaded(cached) }
                return
            }
        }
        try {
            executor.execute {
                val result = find(track)
                synchronized(cache) { cache[track.uri] = result }
                mainHandler.post {
                    if (!closed) callback.loaded(result)
                }
            }
        } catch (_: RejectedExecutionException) {
            // Activity is already closing.
        }
    }

    override fun close() {
        closed = true
        synchronized(cache) { cache.clear() }
        executor.shutdownNow()
    }

    private fun find(track: Track): LrcDocument {
        sidecars(track).forEach { candidate ->
            val value = read(candidate)
            if (value.isNotBlank()) return LrcParser().parse(value)
        }
        try {
            val embedded = EmbeddedLyricsReader().read(context, Uri.parse(track.uri))
            if (embedded.isNotBlank()) return LrcParser().parse(embedded)
        } catch (error: RuntimeException) {
            VoltuneLog.failure("embedded_lyrics_uri_failed", error)
        }
        return LrcDocument(ArrayList(), "", false)
    }

    private fun sidecars(track: Track): List<Uri> {
        val result = ArrayList<Uri>()
        try {
            val source = Uri.parse(track.uri)
            if (
                source.scheme?.equals("content", ignoreCase = true) != true ||
                !DocumentsContract.isDocumentUri(context, source)
            ) {
                return result
            }
            val id = DocumentsContract.getDocumentId(source)
            val slash = id.lastIndexOf('/')
            val parent = if (slash < 0) "" else id.substring(0, slash + 1)
            val filename = if (slash < 0) id else id.substring(slash + 1)
            val extension = filename.lastIndexOf('.')
            val base = if (extension > 0) filename.substring(0, extension) else filename
            addCandidate(result, source, "$parent$base.lrc")
            addCandidate(result, source, "$parent$base.txt")
        } catch (_: RuntimeException) {
            // A malformed or unsupported provider URI simply has no sidecar lyrics.
        }
        return result
    }

    private fun addCandidate(target: MutableList<Uri>, source: Uri, documentId: String) {
        try {
            val candidate = if (source.toString().contains("/tree/")) {
                DocumentsContract.buildDocumentUriUsingTree(source, documentId)
            } else {
                DocumentsContract.buildDocumentUri(source.authority, documentId)
            }
            target.add(candidate)
        } catch (_: RuntimeException) {
            // The provider does not support a sibling document URI.
        }
    }

    private fun read(uri: Uri): String {
        return try {
            context.contentResolver.openInputStream(uri).use { input ->
                if (input == null) return ""
                val output = ByteArrayOutputStream()
                val buffer = ByteArray(BUFFER_BYTES)
                var total = 0
                while (true) {
                    val count = input.read(buffer)
                    if (count < 0) break
                    total += count
                    if (total > MAX_LYRICS_BYTES) return ""
                    output.write(buffer, 0, count)
                }
                val value = output.toString(Charsets.UTF_8.name())
                if (value.startsWith(BYTE_ORDER_MARK)) value.substring(1) else value
            }
        } catch (_: Exception) {
            ""
        }
    }

    private companion object {
        const val MAX_LYRICS_BYTES = 1024 * 1024
        const val MAX_CACHE_ENTRIES = 12
        const val BUFFER_BYTES = 8192
        const val BYTE_ORDER_MARK = "\ufeff"
    }
}
